server/connections.py

```diff
@@ -0,0 +1,112 @@
+from websockets import WebsocketsCommonProtocol
+
+
+class Connections:
+    """
+    Class for managing WebsocketsCommonProtocol instances
+
+    Records a group of WebsocketsCommonProtocol objects in a dict, keyed to
+    usernames. Implements adding and removing users and tracks the most
+    recently added and most recently departed users.
+    """
+
+    def __init__(self, data: dict = None, user_limit: int = None) -> None:
+        """Initialize class"""
+        self.data = {} if data is None else data
+        self._user_limit = user_limit
+        self._latest_departed_user = self._newest_user = None
+
+    def __len__(self) -> int:
+        """Returns number of current users"""
+        return len(self.data)
+
+    def __str__(self) -> str:
+        """Returns string representation of class dict"""
+        return str(self.data)
+
+    def __repr__(self) -> str:
+        """Returns string representation of class dict"""
+        return str(self)
+
+    @property
+    def user_limit(self) -> None | int:
+        """Returns current user limit"""
+        return self._user_limit
+
+    @user_limit.setter
+    def user_limit(self, value: int) -> None:
+        """Sets user limit to a value. Does not perform validation"""
+        self._user_limit = value
+
+    @property
+    def user_count(self) -> int:
+        """
+        User count property
+
+        Returns `_user_count`, which contains the number of users before
+        at the last update
+        """
+        return self._user_count
+
+    def update_user_count(self) -> None:
+        """Updates user count by checking length of the class dict"""
+        self._user_count = len(self)
+
+    @property
+    def latest_departed_user(self) -> None | str:
+        """
+        Latest departed user property
+
+        Returns username of most recently departed user, None if no one has
+        left yet
+        """
+        return self._last_departed_user
+
+    @property
+    def newest_user(self) -> None | str:
+        """
+        Newest user property
+
+        Returns username of most recently joined user,
+        None if no one has joined yet
+        """
+        return self._newest_user
```

Comment:
Please switch this to return a list of users instead, as it makes sense for users who join later than others as they won't have a way to see all the users

Reply:
I'm in the process of making this a PR without the container class at the moment.

Reply:
Not sure I follow. Do you mean just one method to return a list of all users, or a list ordered by join time?

Reply:
You could just replace those two methods with a method that returns this: `str(list(CONNECTIONS.data.keys())).replace("'", '"')

Reply:
Oh, I see, it has to be a string representation of current users. I'll change it.

Reply:
But how would that allow tracking recently entered or exited usernames?

Reply:
when we remove a user from the collection, it is removed from the list. and when a user joins or leaves, we would send a new list.

Reply:
So it's the whole list that should be sent with the event whether the user leaves or joins, meaning tracking individual users is unimportant?

Reply:
Yep. This is the easiest way I've found to keep clients updated on all users.

Reply:
Ah, sorry I misunderstood. I'll make the changes.